connectivity/FEATURE_BLE/include/ble/internal/PalGap.h

```diff
@@ -1879,175 +1547,98 @@ struct Gap {
-        );
-    }
+    ) = 0;
 
     /**
-     * Register a callback which will handle Gap events.
+     * Register a callback which will handle PalGap events.
      *
-     * @param cb The callback object which will handle Gap events from the
+     * @param cb The callback object which will handle PalGap events from the
      * LE subsystem.
      * It accept a single parameter in input: The event received.
      */
-    void when_gap_event_received(mbed::Callback<void(const GapEvent &)> cb)
-    {
-        _gap_event_cb = cb;
-    }
+    virtual void when_gap_event_received(mbed::Callback<void(const GapEvent &)> cb) = 0;
 
-public:
     /**
     * Sets the event handler that us called by the PAL porters to notify the stack of events
     * which will in turn be passed onto the user application when appropriate.
     *
     * @param[in] event_handler the new event handler interface implementation. Memory
     * owned by caller who is responsible for updating this pointer if interface changes.
     */
-    void set_event_handler(EventHandler *event_handler)
-    {
-        _pal_event_handler = event_handler;
-    }
-
-    EventHandler *get_event_handler()
-    {
-        return _pal_event_handler;
-    }
+    virtual void set_event_handler(PalGapEventHandler *event_handler) = 0;
 
-protected:
-    EventHandler *_pal_event_handler;
-
-protected:
-    Gap() : _pal_event_handler(NULL)
-    {
-    }
-
-    ~Gap()
-    {
-    }
-
-    /**
-     * Implementation shall call this function whenever the LE subsystem
-     * generate a Gap event.
-     *
-     * @param gap_event The event to emit to higher layer.
-     */
-    void emit_gap_event(const GapEvent &gap_event)
-    {
-        if (_gap_event_cb) {
-            _gap_event_cb(gap_event);
-        }
-    }
-
-public:
     /**
-     * Create an ALL_PHYS parameter used in LE Set PHY Command
-     * and LE Set Default PHY Command.
-     * @see BLUETOOTH SPECIFICATION Version 5.0 | Vol 2, Part E - 7.8.49
+     * Return the registered event handler.
+     * @return Pointer to event handler
      */
-    static uint8_t create_all_phys_value(
-        const phy_set_t &tx_phys,
-        const phy_set_t &rx_phys
-    )
-    {
-        /* if phy set is empty set corresponding all_phys bit to 1 */
-        uint8_t all_phys = 0;
-        if (tx_phys.value() == 0) {
-            all_phys |= 0x01;
-        }
-        if (rx_phys.value() == 0) {
-            all_phys |= 0x02;
-        }
-        return all_phys;
-    }
-
-private:
-    /**
-     * Callback called when an event is emitted by the LE subsystem.
-     */
-    mbed::Callback<void(const GapEvent &)> _gap_event_cb;
-
-private:
-    // Disallow copy construction and copy assignment.
-    Gap(const Gap &);
-
-    Gap &operator=(const Gap &);
+    virtual PalGapEventHandler *get_event_handler() = 0;
 };
 
-} // namespace pal
+} // namespace interface
 } // namespace ble
 
+
+#if (BLE_PAL_API_IMPLEMENTATION == 1)
+#include "ble/internal/PalGapImpl.h"
```

Comment:
This file is in TARGET_CORDIO - we'd better not include it here.
How to achieve that? In non-Cordio code, just use `ble::interface::PalGap` where possible, and this is the point of virtual functions. It's only a few lines' change and it worked as far as I tried.
Same for the other virtual Pal classes.

Once the generic code has no dependency on headers in TARGET_CORDIO, the headers/classes in the latter can be renamed to contain "Cordio" in my opinion.

Reply:
The point is to remove virtual calls. Alternative implementation can override this include with their implementation.